contracts/cw721-base/src/state.rs

```diff
@@ -1,35 +1,201 @@
+///   - `DefaultOptionalNftExtension`: either with nft metadata (`Some<NftExtension>`) or none `None`.
+///   - `DefaultOptionalCollectionExtension`: either with collection metadata (`Some<CollectionExtension>`) or none `None`.
+///
+/// Example:
+/// ```rust
+/// // instantiate:
+/// let contract = Cw721Contract::<
+///     DefaultOptionalNftExtension, // use `Option<Empty>` or `Empty` for no nft metadata
+///     DefaultOptionalNftExtensionMsg, // use `Option<Empty>` or `Empty` for no nft metadata
+///     DefaultOptionalCollectionExtension, // use `Option<Empty>` or `Empty` for no collection metadata
+///     DefaultOptionalCollectionExtensionMsg, // use `Option<Empty>` or `Empty` for no collection metadata
+///     Empty, // no custom extension msg
+///     Empty, // no custom query msg
+///     Empty, // no custom response msg
+/// >::default();
+/// let info = mock_info(CREATOR, &[]);
+/// let init_msg = Cw721InstantiateMsg {
+///     name: "SpaceShips".to_string(),
+///     symbol: "SPACE".to_string(),
+///     collection_info_extension: None,
+///     minter: None,
+///     creator: None,
+///     withdraw_address: None,
+/// };
+/// //...
+/// // mint:
+/// let token_id = "Enterprise";
+/// let token_uri = Some("https://starships.example.com/Starship/Enterprise.json".into());
+/// let extension = Some(NftExtensionMsg {
+///     description: Some("description1".into()),
+///     name: Some("name1".to_string()),
+///     attributes: Some(vec![Trait {
+///         display_type: None,
+///         trait_type: "type1".to_string(),
+///         value: "value1".to_string(),
+///     }]),
+///     ..NftExtensionMsg::default()
+/// });
+/// let exec_msg = Cw721ExecuteMsg::<
+///     DefaultOptionalNftExtensionMsg,
+///     DefaultOptionalCollectionExtensionMsg,
+///     Empty,
+/// >::Mint {
+///     token_id: token_id.to_string(),
+///     owner: "john".to_string(),
+///     token_uri: token_uri.clone(),
+///     extension: extension.clone(), // use `extension: None` for no metadata
+/// };
+/// //...
+/// ```
 pub struct Cw721Contract<
     'a,
-    // Metadata defined in NftInfo (used for mint).
-    TMetadataExtension,
+    // NftInfo extension (onchain metadata).
+    TNftExtension,
+    // NftInfo extension msg for onchain metadata.
+    TNftExtensionMsg,
+    // CollectionInfo extension (onchain attributes).
+    TCollectionExtension,
+    // CollectionInfo extension msg for onchain collection attributes.
+    TCollectionExtensionMsg,
+    // Custom extension msg for custom contract logic. Default implementation is a no-op.
+    TExtensionMsg,
+    // Custom query msg for custom contract logic. Default implementation returns an empty binary.
+    TExtensionQueryMsg,
     // Defines for `CosmosMsg::Custom<T>` in response. Barely used, so `Empty` can be used.
-    TCustomResponseMessage,
-    // Message passed for updating metadata.
-    TMetadataExtensionMsg,
+    TCustomResponseMsg,
```

Comment:
... last but not least the generic `Cw721Contract` - e.g. for custom nft and collection extensions.